File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer,Integer> mp=new HashMap<>();
        int windowstart=0;
        int maxFruit=0;
        int n=fruits.length;

        for(int i=0;i<n;i++){
            mp.put(fruits[i],mp.getOrDefault(fruits[i],0)+1);

            while(mp.size()>2){
                int leftfruit=fruits[windowstart];
                mp.put(leftfruit,(mp.get(leftfruit)-1));

                if(mp.get(leftfruit)<=0){
                    mp.remove(leftfruit);
                }
                windowstart++;
            }

            maxFruit=Math.max(maxFruit,i-windowstart+1);
        }
        return maxFruit;
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function(fruits) {
    let mp=new Map();
    let windowStart=0;
    let maxFruit=0;

    for(let i=0;i<fruits.length;i++){
        mp.set(fruits[i],(mp.get(fruits[i]) || 0)+1);

        while(mp.size>2){
            let fruit=fruits[windowStart];
            mp.set(fruit,(mp.get(fruit) || 0)-1);

            if(mp.get(fruit)==0)
            {
                mp.delete(fruit)
            }
            windowStart++;
        }
        maxFruit=Math.max(maxFruit,i-windowStart+1);
    }
    return maxFruit;
};